tools/header_order.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python
+
+# Enforce header order in a a given file. This will only reorder in the first sequence of contiguous
+# #include statements, so it will not play well with #ifdef.
+#
+# This attempts to enforce the guidelines at
+# https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes
+# with some allowances for Envoy-specific idioms.
+#
+# There is considerable overlap with what this does and clang-format's IncludeCategories (see
+# https://clang.llvm.org/docs/ClangFormatStyleOptions.html). But, clang-format doesn't seem smart
+# enough to handle block splitting and correctly detecting the main header subject to the Envoy
+# canonical paths.
+
+import re
+import sys
+
+
+def ReorderHeaders(path):
+  with open(path, 'r') as f:
+    source = f.read()
+
+  all_lines = iter(source.split('\n'))
+  before_includes_lines = []
+  includes_lines = []
+  after_includes_lines = []
+
+  try:
+    while True:
+      line = all_lines.next()
+      if line.startswith('#include'):
+        includes_lines.append(line)
+        break
+      before_includes_lines.append(line)
+  except StopIteration:
+    pass
+
+  try:
+    while True:
+      line = all_lines.next()
+      if not line:
+        continue
+      if not line.startswith('#include'):
+        after_includes_lines.append(line)
+        break
+      includes_lines.append(line)
+  except StopIteration:
+    pass
+
+  after_includes_lines += list(all_lines)
+
+  def file_header_filter():
```

Comment:
nit: it took me 60s to figure out that this is figuring out the associated files header, add comment?